doc/doxygen/man/man3/Create_octree.3
.TH "Create_octree" 3 "Wed Oct 6 2021" "MAGRATHEA/PATHFINDER" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Create_octree \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <create_octree\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameter , class Cone , class FileList , class Sphere , typename Integer , typename Scalar > void \fBPreparationHDF5_from_particles\fP (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameter &\fBparameters\fP, const Integer ntasks, const Integer rank, const std::vector< \fBCone\fP > &cone, const std::vector< \fBCone\fP > &coneIfRot, const Scalar thetay, const Scalar thetaz, \fBFileList\fP &conefile, const Sphere &microsphere)"
.br
.RI "\fICreate Octree binary files\&. \fP"
.ti -1c
.RI "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameter , class Cone , class Sphere , class FileList , class Index2 , typename Integer , typename Scalar , class Cosmology > void \fBPreparationBinary\fP (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameter &\fBparameters\fP, const Integer ntasks, const Integer rank, const std::vector< \fBCone\fP > &cone, \fBFileList\fP &conefile, const Sphere &microsphere, Index2 &filetree, const Scalar h, const Scalar omegam, const Scalar lboxmpch, Scalar &amin, const Cosmology &cosmology)"
.br
.RI "\fICreate Octree binary files\&. \fP"
.ti -1c
.RI "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameter , class Cone , class FileList , class Sphere , typename Integer , typename Scalar , class Cosmology > void \fBPreparationASCII\fP (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameter &\fBparameters\fP, const Integer ntasks, const Integer rank, const std::vector< \fBCone\fP > &cone, const std::vector< \fBCone\fP > &coneIfRot, const std::array< std::array< double, 3 >, 3 > &rotm1, \fBFileList\fP &conefile, const Sphere &microsphere, const Scalar h, const Scalar omegam, const Scalar lboxmpch, Scalar &amin, const Cosmology &cosmology)"
.br
.RI "\fICreate Octree binary files\&. \fP"
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "template<class Parameters , class Map > static void \fBReadParamFile\fP (Parameters &\fBparameters\fP, Map &parameter)"
.br
.RI "\fIRead parameter file\&. \fP"
.ti -1c
.RI "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameters , class Cone , class FileList , class Sphere , typename Integer , typename Scalar , class Cosmology > static void \fBPreparationHDF5_from_cells\fP (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &\fBparameters\fP, const Integer ntasks, const Integer rank, const std::vector< \fBCone\fP > &cone, const std::vector< \fBCone\fP > &coneIfRot, const std::array< std::array< double, 3 >, 3 > &rotm1, const Scalar thetay, const Scalar thetaz, \fBFileList\fP &conefile, const Sphere &microsphere, const Scalar h, const Scalar omegam, const Scalar lboxmpch, Scalar &amin, const Cosmology &cosmology)"
.br
.RI "\fICreate Octree binary files\&. \fP"
.ti -1c
.RI "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameters , class Cone , class FileList , class Sphere , typename Integer , typename Scalar > static void \fBPreparationHDF5_from_particles\fP (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &\fBparameters\fP, const Integer ntasks, const Integer rank, const std::vector< \fBCone\fP > &cone, const std::vector< \fBCone\fP > &coneIfRot, const Scalar thetay, const Scalar thetaz, \fBFileList\fP &conefile, const Sphere &microsphere)"
.br
.ti -1c
.RI "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameters , class Cone , class Sphere , class FileList , class Index2 , typename Integer , typename Scalar , class Cosmology > static void \fBPreparationBinary\fP (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &\fBparameters\fP, const Integer ntasks, const Integer rank, const std::vector< \fBCone\fP > &cone, \fBFileList\fP &conefile, const Sphere &microsphere, Index2 &filetree, const Scalar h, const Scalar omegam, const Scalar lboxmpch, Scalar &amin, const Cosmology &cosmology)"
.br
.ti -1c
.RI "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameters , class Cone , class FileList , class Sphere , typename Integer , typename Scalar , class Cosmology > static void \fBPreparationASCII\fP (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &\fBparameters\fP, const Integer ntasks, const Integer rank, const std::vector< \fBCone\fP > &cone, const std::vector< \fBCone\fP > &coneIfRot, const std::array< std::array< double, 3 >, 3 > &rotm1, \fBFileList\fP &conefile, const Sphere &microsphere, const Scalar h, const Scalar omegam, const Scalar lboxmpch, Scalar &amin, const Cosmology &cosmology)"
.br
.ti -1c
.RI "template<class Parameters , typename Type1 , template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container > static void \fBCreateOctreeWithCIC\fP (bool &continue_refining, Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &\fBparameters\fP, std::vector< Type1 > &pos_part, std::vector< Type1 > &force_part, std::vector< Type1 > &potential_part, std::vector< Type1 > &a_part)"
.br
.RI "\fICreate Octree using CIC\&. \fP"
.ti -1c
.RI "template<class Parameters , typename Type1 , template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container > static void \fBCreateOctreeWithTSC\fP (bool &continue_refining, Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &\fBparameters\fP, std::vector< Type1 > &pos_part, std::vector< Type1 > &force_part, std::vector< Type1 > &potential_part, std::vector< Type1 > &a_part)"
.br
.RI "\fICreate Octree using TSC\&. \fP"
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "template<class Parameters , typename Type1 , template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container > void Create_octree::CreateOctreeWithCIC (bool &continue_refining, Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &parameters, std::vector< Type1 > &pos_part, std::vector< Type1 > &force_part, std::vector< Type1 > &potential_part, std::vector< Type1 > &a_part)\fC [static]\fP"

.PP
Create Octree using CIC\&. Create Octree using CIC from particles 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIParameters\fP Parameters type 
.br
\fIType1\fP Type1 type 
.br
\fIOctree\fP octree type 
.br
\fIType\fP type type 
.br
\fIIndex\fP index type 
.br
\fIData\fP data type 
.br
\fIDimension\fP Number of dimensions 
.br
\fIPosition\fP position type 
.br
\fIExtent\fP extent type 
.br
\fIElement\fP element type 
.br
\fIContainer\fP container type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIcontinue_refining\fP Bolean which controls if we continue refinement 
.br
\fIoctree\fP Octree be to filled using particles 
.br
\fIparameters\fP Parameters structure 
.br
\fIpos_part\fP Position of particles 
.br
\fIforce_part\fP Force of particles 
.br
\fIpotential_part\fP Potential of particles 
.br
\fIa_part\fP Scale factor of particles 
.RE
.PP

.SS "template<class Parameters , typename Type1 , template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container > void Create_octree::CreateOctreeWithTSC (bool &continue_refining, Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &parameters, std::vector< Type1 > &pos_part, std::vector< Type1 > &force_part, std::vector< Type1 > &potential_part, std::vector< Type1 > &a_part)\fC [static]\fP"

.PP
Create Octree using TSC\&. Create Octree using TSC 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIParameters\fP Parameters type 
.br
\fIType1\fP Type1 type 
.br
\fIOctree\fP octree type 
.br
\fIType\fP type type 
.br
\fIIndex\fP index type 
.br
\fIData\fP data type 
.br
\fIDimension\fP Number of dimensions 
.br
\fIPosition\fP position type 
.br
\fIExtent\fP extent type 
.br
\fIElement\fP element type 
.br
\fIContainer\fP container type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIcontinue_refining\fP Bolean which controls if we continue refinement 
.br
\fIoctree\fP Octree be to filled using particles 
.br
\fIparameters\fP Parameters structure 
.br
\fIpos_part\fP Position of particles 
.br
\fIforce_part\fP Force of particles 
.br
\fIpotential_part\fP Potential of particles 
.br
\fIa_part\fP Scale factor of particles 
.RE
.PP

.SS "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameters , class Cone , class FileList , class Sphere , typename Integer , typename Scalar , class Cosmology > static void Create_octree::PreparationASCII (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &parameters, const Integerntasks, const Integerrank, const std::vector< \fBCone\fP > &cone, const std::vector< \fBCone\fP > &coneIfRot, const std::array< std::array< double, 3 >, 3 > &rotm1, \fBFileList\fP &conefile, const Sphere &microsphere, const Scalarh, const Scalaromegam, const Scalarlboxmpch, Scalar &amin, const Cosmology &cosmology)\fC [static]\fP"

.SS "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameter , class Cone , class FileList , class Sphere , typename Integer , typename Scalar , class Cosmology > void Create_octree::PreparationASCII (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameter &parameters, const Integerntasks, const Integerrank, const std::vector< \fBCone\fP > &cone, const std::vector< \fBCone\fP > &coneIfRot, const std::array< std::array< double, 3 >, 3 > &rotm1, \fBFileList\fP &conefile, const Sphere &microsphere, const Scalarh, const Scalaromegam, const Scalarlboxmpch, Scalar &amin, const Cosmology &cosmology)"

.PP
Create Octree binary files\&. Create Octree binary files from ASCII iles\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIOctree\fP octree type 
.br
\fIType\fP type type 
.br
\fIIndex\fP index type 
.br
\fIData\fP data type 
.br
\fIDimension\fP Number of dimensions 
.br
\fIPosition\fP position type 
.br
\fIExtent\fP extent type 
.br
\fIElement\fP element type 
.br
\fIContainer\fP container type 
.br
\fIParameters\fP Parameters type 
.br
\fI\fBCone\fP\fP cone type 
.br
\fIFilelist\fP filelist type 
.br
\fISphere\fP sphere type 
.br
\fIInteger\fP Integer type 
.br
\fIScalar\fP scalar type 
.br
\fICosmology\fP cosmology type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIoctree\fP Octree to be filled 
.br
\fIparameters\fP Parameters structure 
.br
\fIntasks\fP Number of tasks 
.br
\fIrank\fP Process rank 
.br
\fIcone\fP \fBCone\fP properties 
.br
\fIconeIfRot\fP \fBCone\fP with possibly rotation 
.br
\fIrotm1\fP Rotation matrix for narrow cells 
.br
\fIconefile\fP \fBCone\fP names in conedir 
.br
\fImicrosphere\fP Central buffer zone for Octree 
.br
\fIh\fP dimensionless Hubble parameter 
.br
\fIomegam\fP Matter density fraction 
.br
\fIlboxmpch\fP Size of simulation box 
.br
\fIamin\fP minimum value scale factor 
.br
\fIcosmology\fP Cosmological tables 
.RE
.PP

.SS "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameters , class Cone , class Sphere , class FileList , class Index2 , typename Integer , typename Scalar , class Cosmology > static void Create_octree::PreparationBinary (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &parameters, const Integerntasks, const Integerrank, const std::vector< \fBCone\fP > &cone, \fBFileList\fP &conefile, const Sphere &microsphere, Index2 &filetree, const Scalarh, const Scalaromegam, const Scalarlboxmpch, Scalar &amin, const Cosmology &cosmology)\fC [static]\fP"

.SS "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameter , class Cone , class Sphere , class FileList , class Index2 , typename Integer , typename Scalar , class Cosmology > void Create_octree::PreparationBinary (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameter &parameters, const Integerntasks, const Integerrank, const std::vector< \fBCone\fP > &cone, \fBFileList\fP &conefile, const Sphere &microsphere, Index2 &filetree, const Scalarh, const Scalaromegam, const Scalarlboxmpch, Scalar &amin, const Cosmology &cosmology)"

.PP
Create Octree binary files\&. Create Octree binary files from Binary\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIOctree\fP octree type 
.br
\fIType\fP type type 
.br
\fIIndex\fP index type 
.br
\fIData\fP data type 
.br
\fIDimension\fP Number of dimensions 
.br
\fIPosition\fP position type 
.br
\fIExtent\fP extent type 
.br
\fIElement\fP element type 
.br
\fIContainer\fP container type 
.br
\fIParameters\fP Parameters type 
.br
\fI\fBCone\fP\fP cone type 
.br
\fIFilelist\fP filelist type 
.br
\fISphere\fP sphere type 
.br
\fIInteger\fP Integer type 
.br
\fIScalar\fP scalar type 
.br
\fICosmology\fP cosmology type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIoctree\fP Octree to be filled 
.br
\fIparameters\fP Parameters structure 
.br
\fIntasks\fP Number of tasks 
.br
\fIrank\fP Process rank 
.br
\fIcone\fP \fBCone\fP properties 
.br
\fIconefile\fP \fBCone\fP names in conedir 
.br
\fImicrosphere\fP Central buffer zone for Octree 
.br
\fIfiletree\fP Type of Octree containing strings 
.br
\fIh\fP dimensionless Hubble parameter 
.br
\fIomegam\fP Matter density fraction 
.br
\fIlboxmpch\fP Size of simulation box 
.br
\fIamin\fP minimum value scale factor 
.br
\fIcosmology\fP Cosmological tables 
.RE
.PP

.SS "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameters , class Cone , class FileList , class Sphere , typename Integer , typename Scalar , class Cosmology > void Create_octree::PreparationHDF5_from_cells (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &parameters, const Integerntasks, const Integerrank, const std::vector< \fBCone\fP > &cone, const std::vector< \fBCone\fP > &coneIfRot, const std::array< std::array< double, 3 >, 3 > &rotm1, const Scalarthetay, const Scalarthetaz, \fBFileList\fP &conefile, const Sphere &microsphere, const Scalarh, const Scalaromegam, const Scalarlboxmpch, Scalar &amin, const Cosmology &cosmology)\fC [static]\fP"

.PP
Create Octree binary files\&. Create Octree binary files from HDF5\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIOctree\fP octree type 
.br
\fIType\fP type type 
.br
\fIIndex\fP index type 
.br
\fIData\fP data type 
.br
\fIDimension\fP Number of dimensions 
.br
\fIPosition\fP position type 
.br
\fIExtent\fP extent type 
.br
\fIElement\fP element type 
.br
\fIContainer\fP container type 
.br
\fIParameters\fP Parameters type 
.br
\fI\fBCone\fP\fP cone type 
.br
\fIFilelist\fP filelist type 
.br
\fISphere\fP sphere type 
.br
\fIInteger\fP Integer type 
.br
\fIScalar\fP scalar type 
.br
\fICosmology\fP cosmology type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIoctree\fP Octree to be filled 
.br
\fIparameters\fP Parameters structure 
.br
\fIntasks\fP Number of tasks 
.br
\fIrank\fP Process rank 
.br
\fIcone\fP \fBCone\fP properties 
.br
\fIconeIfRot\fP \fBCone\fP with possibly rotation 
.br
\fIrotm1\fP Rotation matrix for narrow cells 
.br
\fIthetay\fP Semi-angle for solid angle in direction y 
.br
\fIthetaz\fP Semi-angle for solid angle in direction z 
.br
\fIconefile\fP \fBCone\fP names in conedir 
.br
\fImicrosphere\fP Central buffer zone for Octree 
.br
\fIh\fP Dimensionless Hubble parameter 
.br
\fIomegam\fP Matter density fraction 
.br
\fIlboxmpch\fP Size of simulation box 
.br
\fIamin\fP minimum value scale factor 
.br
\fIcosmology\fP Cosmological tables 
.RE
.PP

.SS "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameters , class Cone , class FileList , class Sphere , typename Integer , typename Scalar > static void Create_octree::PreparationHDF5_from_particles (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameters &parameters, const Integerntasks, const Integerrank, const std::vector< \fBCone\fP > &cone, const std::vector< \fBCone\fP > &coneIfRot, const Scalarthetay, const Scalarthetaz, \fBFileList\fP &conefile, const Sphere &microsphere)\fC [static]\fP"

.SS "template<template< typename Type, class Index, class Data, unsigned int Dimension, class Position, class Extent, class Element, class Container > class Octree, typename Type , class Index , class Data , unsigned int Dimension, class Position , class Extent , class Element , class Container , class Parameter , class Cone , class FileList , class Sphere , typename Integer , typename Scalar > void Create_octree::PreparationHDF5_from_particles (Octree< Type, Index, Data, Dimension, Position, Extent, Element, Container > &octree, const Parameter &parameters, const Integerntasks, const Integerrank, const std::vector< \fBCone\fP > &cone, const std::vector< \fBCone\fP > &coneIfRot, const Scalarthetay, const Scalarthetaz, \fBFileList\fP &conefile, const Sphere &microsphere)"

.PP
Create Octree binary files\&. Create Octree binary files from HDF5\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIOctree\fP octree type 
.br
\fIType\fP type type 
.br
\fIIndex\fP index type 
.br
\fIData\fP data type 
.br
\fIDimension\fP Number of dimensions 
.br
\fIPosition\fP position type 
.br
\fIExtent\fP extent type 
.br
\fIElement\fP element type 
.br
\fIContainer\fP container type 
.br
\fIParameters\fP Parameters type 
.br
\fI\fBCone\fP\fP cone type 
.br
\fIFilelist\fP filelist type 
.br
\fISphere\fP sphere type 
.br
\fIInteger\fP Integer type 
.br
\fIScalar\fP scalar type 
.br
\fICosmology\fP cosmology type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIoctree\fP Octree to be filled 
.br
\fIparameters\fP Parameters structure 
.br
\fIntasks\fP Number of tasks 
.br
\fIrank\fP Process rank 
.br
\fIcone\fP \fBCone\fP properties 
.br
\fIconeIfRot\fP \fBCone\fP with possibly rotation 
.br
\fIconefile\fP \fBCone\fP names in conedir 
.br
\fImicrosphere\fP Central buffer zone for Octree 
.br
\fIh\fP Dimensionless Hubble parameter 
.br
\fIomegam\fP Matter density fraction 
.br
\fIlboxmpch\fP Size of simulation box 
.br
\fIamin\fP minimum value scale factor 
.br
\fIcosmology\fP Cosmological tables 
.RE
.PP

.SS "template<class Parameters , class Map > void Create_octree::ReadParamFile (Parameters &parameters, Map &parameter)\fC [static]\fP"

.PP
Read parameter file\&. Read and put in a structure the parameters\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIParameters\fP structure type 
.br
\fIMap\fP map type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIparameters\fP Structure containing the parameters\&. 
.br
\fIparameter\fP Contains parameters to be rewritten 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for MAGRATHEA/PATHFINDER from the source code\&.


src/create_octree.cpp
/* ******************************* CREATE_OCTREE ******************************** */
/*////////////////////////////////////////////////////////////////////////////*/
// PROJECT :        PATHFINDER
// TITLE :          Create_octree
// DESCRIPTION :    Main function of the raytracer
// AUTHOR(S) :      Vincent Reverdy (vince.rev@gmail.com), Michel-Andrès Breton (michel-andres.breton@obspm.fr)
// CONTRIBUTIONS :  [Vincent Reverdy (2012-2013), Michel-Andrès Breton (2015-2021)]
// LICENSE :        CECILL-B License
/*////////////////////////////////////////////////////////////////////////////*/
/// \file           raytracer.cpp
/// \brief          Main function of the raytracer
/// \author         Vincent Reverdy (vince.rev@gmail.com), Michel-Andrès Breton (michel-andres.breton@obspm.fr)
/// \date           2012-2013, 2015-2021
/// \copyright      CECILL-B License
/*////////////////////////////////////////////////////////////////////////////*/



// ------------------------------ PREPROCESSOR ------------------------------ //
#include <ctime>
// Include C++
#include <iostream>
#include <iomanip>
#include <fstream>
#include <algorithm>
#include <utility>
#include <random>
#include <string>
#include <vector>
#include <deque>
#include <array>
#include <tuple>
#include <atomic>
#include <mutex>
// Include libs
#include <mpi.h>
// Include project
#include "magrathea/simplehyperoctree.h"
#include "magrathea/simplehyperoctreeindex.h"
#include "magrathea/hypersphere.h"
#include "magrathea/hypercube.h"
#include "magrathea/constants.h"
#include "magrathea/evolution.h"
#include "magrathea/timer.h"
#include "cone.h"
#include "utility.h"
#include "input.h"
#include "output.h"
#include "integrator.h"
#include "miscellaneous.h"
#include "create_octree.h"
// Octree
#ifdef VELOCITYFIELD
#include "gravity2.h" // with velocity slots
#else
#include "gravity.h"
#endif
// Include C++ HDF5 project
#include "TReadHDF5.h"


// Misc
using namespace magrathea;
// -------------------------------------------------------------------------- //



// ---------------------------------- MAIN ---------------------------------- //
// Main function
/// \brief          Main function.
/// \details        Main function of the program. It takes a namelist as a 
///                 parameter and run the raytracing using MPI parallelization.
///                 To compile it, use make
///                 To run it, use:
///                 mpirun -np NTASKS ./raytracer raytracer.txt
/// \param[in]      argc Number of arguments.
/// \param[in]      argv List of arguments.
/// \return         Zero on success, error code otherwise.
int main(int argc, char* argv[])
{
    // Constants

    using integer = int;
    using uint = unsigned int;
    using real = double;
    using floating = float;
    using point = std::array<real, 3>;
    using position = std::ratio<0>;
    using extent = std::ratio<EXTENT>;
    using indexing = __uint128_t;
    static constexpr uint zero = 0;
    static constexpr uint one = 1;
    static constexpr real rone = 1.; 
    static constexpr uint two = 2;
    static constexpr uint dimension = 3;
    static constexpr uint nreference = 5; // Used to set homogeneous octree
    static constexpr real rposition = static_cast<real>(position::num)/static_cast<real>(position::den);
    static constexpr point center({{rposition, rposition, rposition}});
    static constexpr real diameter = static_cast<real>(extent::num)/static_cast<real>(extent::den);
    static const std::string namelist = argc > 1 ? std::string(argv[1]) : std::string("raytracer.txt"); 

    // Parameters
    std::map<std::string, std::string> parameter;
    integer nthreads = std::thread::hardware_concurrency();
    integer ntasks = nthreads*zero;
    integer rank = zero;

    // Message passing interface
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &ntasks);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    // Read parameter file
    Miscellaneous::TicketizeFunction(rank, ntasks, [=, &parameter]{parameter = Input::parse(namelist);});
    // Convert strings and put it in struct
    Create_octree::ReadParamFile(parameters, parameter);
    // Initialization
    FileList conefile(parameters.conefmt, zero, parameters.ncones, zero, parameters.conedir);
    SimpleHyperOctree<real, SimpleHyperOctreeIndex<uint, dimension>, std::string, dimension, position, extent> filetree;
    SimpleHyperOctree<real, SimpleHyperOctreeIndex<indexing, dimension>, Gravity<floating, dimension>, dimension, position, extent> octree;
    HyperSphere<dimension, point> sphere(center, diameter/two);
    HyperSphere<dimension, point> microsphere(center, rone/parameters.microcoeff);
    std::vector<Cone<point> > cone(parameters.ncones);
    std::vector<Cone<point> > coneIfRot(parameters.ncones);
    std::array<std::vector<real>, two+two> cosmology;
    Evolution<Photon<real, dimension> > reference;
    real h = zero;
    real omegam = zero;
    real lboxmpch = zero;
    real amin = zero;
    real thetay(0), thetaz(0);
    std::array< std::array< double, 3 >, 3 > rotm1 = {{zero}};
    const point vobs0 = {0,0,0}; // No peculiar velocity for homogeneous quantities
 
    if(rank == 0) 
	std::cout<<"#### MAGRATHEA_PATHFINDER "<<std::endl; 
    // Generate cones
    if(parameters.isfullsky){
	if(rank == 0){
	    std::cout<<"## Fullsky cone"<<std::endl;
	}
        Miscellaneous::GenerateFullskyCones(parameters.ncones, cone, coneIfRot, sphere);
    } else{
	if(rank == 0){
	    std::cout<<"## Narrow cone"<<std::endl;
	}
        Miscellaneous::GenerateNarrowCones(parameters, cone, coneIfRot, sphere, rotm1, thetay, thetaz);
    }
    // Read cosmology
    cosmology = Input::acquire(parameters, h, omegam, lboxmpch);
    if(rank == 0){
        std::cout<<"## Parameter file "<<parameters.paramfile<<std::endl;
        std::cout<<"## Evolution file "<<parameters.evolfile<<std::endl;
        std::cout<<"## Cosmology : h = "<<h<<", omega_m = "<<omegam<<", rhoch2 = "<<parameters.rhoch2<<std::endl;
        std::cout<<"## Box length : "<<lboxmpch<<" Mpc/h, 1 Mpc = "<<parameters.mpc<<" meters "<<std::endl;
        std::cout<<"## Coarse level : "<<parameters.ncoarse<<", number of cones = "<<parameters.ncones<<std::endl;
    }
    const double length = lboxmpch*parameters.mpc/h;

    // Construct homogeneous tree 

    Input::homogenize(octree.assign(nreference, zero));
    reference.append(Integrator::launch(center[zero], center[one], center[two], center[zero]+diameter/two, center[one], center[two]));

    // Propagate a photon in a homogeneous cosmology

    Integrator::integrate<-1>(reference, cosmology, octree, vobs0, length, EXTENT*std::pow(two, static_cast<uint>(std::log2(std::get<0>(cosmology).size()/std::pow(two, nreference)+one)+one)+one));
    cosmology = Input::correct(cosmology, reference);
    reference.fullclear();
    octree.fullclear();
   
    // Construct octree

    if (parameters.typefile == 2) { // ASCII
#ifdef VERBOSE
        if(rank == 0) 
	    std::cout<<"# Preparation mode : read ASCII files"<<std::endl;
#endif
	Create_octree::PreparationASCII(octree, parameters, ntasks, rank, cone, coneIfRot, rotm1, conefile, microsphere, h,  omegam, lboxmpch, amin, cosmology);
    } else if (parameters.typefile == 1) { // HDF5
#ifdef VERBOSE
        if(rank == 0) 
	    std::cout<<"# Preparation mode : read HDF5 files"<<std::endl;
#endif
	if(parameters.inputtype == "cells"){
	    Create_octree::PreparationHDF5_from_cells(octree, parameters, ntasks, rank, cone, coneIfRot, rotm1, thetay, thetaz, conefile, microsphere, h,  omegam, lboxmpch, amin, cosmology);
	} else if (parameters.inputtype == "particles"){ 
	    Create_octree::PreparationHDF5_from_particles(octree, parameters, ntasks, rank, cone, coneIfRot, thetay, thetaz, conefile, microsphere);
	} else{
	    std::cout<<"# Please choose 'cells' or 'particles' for inputtype"<<std::endl;
	    std::cout<<"# Error at file "<<__FILE__<<", line : "<<__LINE__<<std::endl;
	    std::terminate();
	}
    } else if (parameters.typefile == 0) { // Binary
#ifdef VERBOSE
        if(rank == 0) 
	    std::cout<<"# Preparation mode : read Binary files"<<std::endl;
#endif
	Create_octree::PreparationBinary(octree, parameters, ntasks, rank, cone, conefile, microsphere, filetree, h,  omegam, lboxmpch, amin, cosmology);
    } else{
        if(rank == 0) {
	    std::cout<<"# Please choose Preparation or Propagation mode"<<std::endl;
	    std::cout<<"# Error at file "<<__FILE__<<", line : "<<__LINE__<<std::endl;
	    std::terminate();
	}
    }

    // Finalization
    MPI_Barrier(MPI_COMM_WORLD);
    MPI_Finalize();
    if(rank == 0)
        std::cout<<"# Run completed !"<<std::endl;
    return 0;
}
// -------------------------------------------------------------------------- //



/*////////////////////////////////////////////////////////////////////////////*/
